### PR TITLE
ARROW-16114: [Docs][Python] Document Parquet FileMetaData

### PR DESCRIPTION
**Pull request by wjones127**
Recently was using this class in a third-party library and had a hard time finding information about these classes.

### CODE REVIEW

File: python/pyarrow/_parquet.pyx [hunk 1]
```diff
@@ -49,6 +49,8 @@ cimport cpython as cp
 
 
 cdef class Statistics(_Weakrefable):
+    """Statistics for a single column in a single row group"""
```
Comment (lidavidm) [inline python/pyarrow/_parquet.pyx:52]:
```suggestion
    """Statistics for a single column in a single row group."""
```

File: python/pyarrow/_parquet.pyx [hunk 2]
```diff
@@ -96,72 +98,87 @@ cdef class Statistics(_Weakrefable):
 
     @property
     def has_min_max(self):
+        """bool: whether min and max are present"""
```
Comment (lidavidm) [inline python/pyarrow/_parquet.pyx:101]: Huh, did not realize numpydoc recommended this style for property docstrings. It looks like the sentences should still end in a period, though?

File: python/pyarrow/_parquet.pyx [hunk 3]
```diff
@@ -96,72 +98,87 @@ cdef class Statistics(_Weakrefable):
...
 
     @property
     def converted_type(self):
+        """str or None: legacy converted type"""
         raw_converted_type = self.statistics.get().descr().converted_type()
         return converted_type_name_from_enum(raw_converted_type)
 
 
 cdef class ParquetLogicalType(_Weakrefable):
+    """Logical type of parquet type"""
```
Comment (lidavidm) [inline python/pyarrow/_parquet.pyx:181]:
```suggestion
    """Logical type of Parquet type."""
```

File: python/pyarrow/_parquet.pyx [hunk 4]
```diff
@@ -292,6 +309,8 @@ cdef _box_flba(ParquetFLBA val, uint32_t len):
 
 
 cdef class ColumnChunkMetaData(_Weakrefable):
+    """Column metadata for a single row group"""
```
Comment (lidavidm) [inline python/pyarrow/_parquet.pyx:312]:
```suggestion
    """Column metadata for a single row group."""
```

File: python/pyarrow/_parquet.pyx [hunk 5]
```diff
@@ -392,45 +418,62 @@ cdef class ColumnChunkMetaData(_Weakrefable):
...
         return self.metadata.total_compressed_size()
 
     @property
     def total_uncompressed_size(self):
+        """int: uncompressed size in bytes"""
         return self.metadata.total_uncompressed_size()
 
 
 cdef class RowGroupMetaData(_Weakrefable):
+    """Metadata for a single row group"""
```
Comment (lidavidm) [inline python/pyarrow/_parquet.pyx:475]:
```suggestion
    """Metadata for a single row group."""
```

File: python/pyarrow/_parquet.pyx [hunk 6]
```diff
@@ -504,6 +550,8 @@ def _reconstruct_filemetadata(Buffer serialized):
 
 
 cdef class FileMetaData(_Weakrefable):
+    """Parquet metadata for a single file"""
```
Comment (lidavidm) [inline python/pyarrow/_parquet.pyx:553]:
```suggestion
    """Parquet metadata for a single file."""
```

File: python/pyarrow/_parquet.pyx [hunk 7]
```diff
@@ -609,6 +672,17 @@ cdef class FileMetaData(_Weakrefable):
             return None
 
     def row_group(self, int i):
+        """Get metadata for row group at index i
```
Comment (lidavidm) [inline python/pyarrow/_parquet.pyx:675]:
```suggestion
        """Get metadata for row group at index i.
```

File: python/pyarrow/_parquet.pyx [hunk 8]
```diff
@@ -670,6 +744,8 @@ cdef class FileMetaData(_Weakrefable):
 
 
 cdef class ParquetSchema(_Weakrefable):
+    """A Parquet schema"""
```
Comment (lidavidm) [inline python/pyarrow/_parquet.pyx:747]:
```suggestion
    """A Parquet schema."""
```

File: python/pyarrow/_parquet.pyx [hunk 9]
```diff
@@ -732,13 +809,25 @@ cdef class ParquetSchema(_Weakrefable):
...
+        column_schema : ColumnSchema
+        """
         if i < 0 or i >= len(self):
             raise IndexError('{0} out of bounds'.format(i))
 
         return ColumnSchema(self, i)
 
 
 cdef class ColumnSchema(_Weakrefable):
+    """Schema for a single column"""
```
Comment (lidavidm) [inline python/pyarrow/_parquet.pyx:830]:
```suggestion
    """Schema for a single column."""
```

File: python/pyarrow/_parquet.pyx [hunk 10]
```diff
@@ -481,14 +524,17 @@ cdef class RowGroupMetaData(_Weakrefable):
```
Comment (lidavidm) [inline python/pyarrow/_parquet.pyx:524]: There are some other methods above that aren't documented, should we document those?

**Review by lidavidm — COMMENTED** [review]
Thanks for adding this!

File: python/pyarrow/_parquet.pyx (hunk 2 above)
Comment (wjones127) [reply python/pyarrow/_parquet.pyx:101]: Yeah I don't think it's adopted anywhere else in Arrow, but it seemed like the best way to do it to me. Using a Return section felt verbose and IIRC doesn't show the type in the attribute summary table.

**Comment by wjones127** [comment]
I've documented the remaining methods and fixed capitalization and missing periods.

File: python/pyarrow/_parquet.pyx (hunk 2 above)
Comment (jorisvandenbossche) [reply python/pyarrow/_parquet.pyx]: Maybe doing something like `Whether min and max are present (bool).` is also an option? 

(I wanted to add the same comment about this style not being consistent with numpydoc)

File: python/pyarrow/_parquet.pyx [hunk 11]
```diff
@@ -75,6 +77,13 @@ cdef class Statistics(_Weakrefable):
                                         self.converted_type)
 
     def to_dict(self):
+        """Get dictionary represenation of statistics.
```
Comment (jorisvandenbossche) [inline python/pyarrow/_parquet.pyx:80]:
```suggestion
        """
        Get dictionary represenation of statistics.
```

Another style nit .., but let's maybe try to do this consistently. I think we mostly start on the next line for multi-line docstring, and below in this diff it's a mix of both.

File: python/pyarrow/_parquet.pyx (hunk 2 above)
Comment (lidavidm) [reply python/pyarrow/_parquet.pyx]: I actually mean that as-is, this seems to be the recommended numpydoc style, at least as parsed by Napoleon: https://github.com/sphinx-contrib/napoleon/blob/dce30797b7a229ccebda4030f65482d501427794/docs/source/example_numpy.py#L280-L282

It just looks quite different from the rest of numpydoc style.

Comment (wjones127) [reply python/pyarrow/_parquet.pyx]: What Joris suggested is also used by some projects: https://developer.lsst.io/python/numpydoc.html#documenting-class-properties

I'm good with either way. Do we prefer putting the type at the end?

Comment (lidavidm) [reply python/pyarrow/_parquet.pyx]: Oh huh.

I think what Joris suggests looks a bit better, but does numpydoc/napoleon parse it properly? If not I'd rather stick with how it is.

Comment (wjones127) [reply python/pyarrow/_parquet.pyx]: > does numpydoc/napoleon parse it properly?

Either version isn't really parsed into anything special. Hence why I have to have the `` :class:`XXX` `` to get cross references.

Comment (lidavidm) [reply python/pyarrow/_parquet.pyx]: In that case I slightly prefer what Joris has but how it is is also fine (so long as it's punctuated)

File: python/pyarrow/_parquet.pyx [hunk 12]
```diff
@@ -174,14 +218,27 @@ cdef class ParquetLogicalType(_Weakrefable):
...
         self.type = type
 
+    def __repr__(self):
+        return "{}\n  {}".format(object.__repr__(self), str(self))
+
     def __str__(self):
         return frombytes(self.type.get().ToString(), safe=True)
 
     def to_json(self):
+        """Get a JSON string containing type and type parameters
```
Comment (lidavidm) [inline python/pyarrow/_parquet.pyx:228]:
```suggestion
        """Get a JSON string containing type and type parameters.
```

**Review by lidavidm — APPROVED** [review]
LGTM, though I agree with Joris we should try to keep things consistently formatted.

**Review by lidavidm — APPROVED** [review]

**Pull request closed without merging**